### DISCUSSION TITLE
Native ReID Feature Extraction for NMS-Free Architectures (YOLO26): Challenges and Potential Solutions

### DISCUSSION THREAD

**Discussion by tejus09** [ultralytics/ultralytics#23400] [Q&A]
# Native ReID Feature Extraction for NMS-Free Architectures (YOLO26): Challenges and Potential Solutions

## Problem Statement

Native reID feature extraction works reliably for NMS-based detectors (YOLOv8, YOLO11) but performs poorly for NMS-free architectures like YOLO26 (`end2end=True`, `reg_max=1`). The core issue lies in how to map final detections back to the feature map locations that produced them.

### NMS-Based Detectors (Working Approach)

For YOLOv8/v11, the implementation in `ultralytics/models/yolo/detect/predict.py` uses:

```python
def get_obj_feats(feat_maps, idxs):
    """Extract object features from the feature maps."""
    s = min(x.shape[1] for x in feat_maps)  # find shortest vector length
    obj_feats = torch.cat(
        [x.permute(0, 2, 3, 1).reshape(x.shape[0], -1, s, x.shape[1] // s).mean(dim=-1) 
         for x in feat_maps], dim=1
    )  # mean reduce all vectors to same length
    return [feats[idx] if idx.shape[0] else [] for feats, idx in zip(obj_feats, idxs)]
```

**Why this works:**
1. NMS returns indices (`return_idxs=True`) that directly indicate which feature map locations produced each detection
2. These indices map to the concatenated feature tensor from all scales (P3, P4, P5)
3. Features are extracted from the **exact locations** that generated the detections

This approach is successfully used in BotSORT tracking (`ultralytics/trackers/bot_sort.py:188-189`):

```python
self.encoder = (
    (lambda feats, s: [f.cpu().numpy() for f in feats])  # native features
    if args.with_reid and self.args.model == "auto"
    ...
)
```

### NMS-Free Detectors (Current Challenge)

For YOLO26 (`end2end=True`), the detection head's `postprocess()` method (in `ultralytics/nn/modules/head.py:203-217`) selects top-k detections via `get_topk_index()`, but:

1. `non_max_suppression()` doesn't return indices for end2end models (see `ultralytics/utils/nms.py:66-70`)
2. The final boxes are already post-processed, losing the mapping to source feature map locations
3. We must **retroactively determine** which feature map locations correspond to each detection

### Approach 1: Multi-Scale Center Point Sampling

Sample features from **all scales** (P3, P4, P5) at the box center, normalize channels, and concatenate:

```python
def get_obj_feats_from_boxes(feat_maps, preds, img_shape):
    # For each detection, sample at box center from all scales
    for k, fm in enumerate(feat_maps):
        gx = (box_center_x / stride_x).round().long()
        gy = (box_center_y / stride_y).round().long()
        v = fm[b, :, gy, gx]  # Sample single point
        # Normalize and concatenate across scales
```

**Results:** Poor discriminability. All detections show high similarity (0.64-0.86), indicating features are not distinctive enough.

### Approach 2: Size-Based Scale Selection

Select a **single best scale** per detection based on object size (mimicking FPN assignment):

```python
# Compute box size in cells at each scale
cells = box_size / stride
# Choose scale where cells is closest to target (e.g., 8 cells)
best_scale = argmin(abs(cells - target_cells))
# Sample only from best_scale
```

**Results:** Worse than Approach 1. Single-scale features lack multi-scale context.

### YOLO11n (NMS-based) - Working ✅

**Detections:**
<img width="869" height="1184" alt="detections_with_ids_11n" src="https://github.com/user-attachments/assets/65c8f65e-8615-4b93-892c-e1531859e8c7" />

**Similarity Matrix:**
<img width="2381" height="1028" alt="similarity_matrix-11n" src="https://github.com/user-attachments/assets/28996c17-724b-4a5e-818c-5ffef2f62b3a" />

- Shows reasonable similarity values (0.12-0.70 range)
- Distinct objects have lower similarity, same objects have higher similarity

### YOLO26n (NMS-free) - Problematic ❌

**Detections:**
<img width="869" height="1184" alt="detections_with_ids_26n" src="https://github.com/user-attachments/assets/d96ec05a-770b-4d7a-9879-c1b2167a22f9" />

**Similarity Matrix:**
<img width="2381" height="1028" alt="similarity_matrix-26n" src="https://github.com/user-attachments/assets/c362e6ba-b1dd-4e7b-ba2f-823b5ce1ccd5" />

- All similarities are high (0.64-0.86 range)
- Features lack discriminability between different objects

## Root Cause Analysis

1. **Loss of spatial correspondence:** Without NMS indices, we don't know which feature map locations produced each detection
2. **Single-point sampling:** Sampling only at box centers misses spatial context within the box
3. **Scale ambiguity:** For end2end models, detections may come from any scale, making scale selection unreliable

## Questions for Maintainers

1. **Is there a way to access the feature map indices** used by the `Detect` head's `postprocess()` method for end2end models? The `get_topk_index()` method returns `idx`, but this is internal to the head.

2. **Would ROI pooling or adaptive pooling** within the box region be a better approach than point sampling? This would capture spatial context similar to how detection heads process features.

3. **Are there architectural differences** in how YOLO26 processes multi-scale features compared to YOLOv8/v11 that we should account for?

4. **Is there a recommended approach** for extracting native reID features from end2end models, or should we use a separate ReID model for these architectures?

## Potential Solutions to Explore

1. **Use ROI pooling/adaptive pooling** within each box region instead of point sampling
2. **Extract features from backbone outputs** (before the head) using the box coordinates

## Environment

- **Ultralytics version:** Latest (from main branch)
- **Models tested:** YOLO11n (NMS-based), YOLO26n (NMS-free)
- **Task:** Native reID feature extraction for object tracking

---

Any guidance on the recommended approach for NMS-free architectures would be greatly appreciated. Thank you for maintaining this excellent library!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tejus09, thank you for your detailed discussion and for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This is an automated response 🤖—an Ultralytics engineer will also jump in to help soon ✅

If this is a 🐛 Bug Report (e.g., you believe native ReID feature extraction behavior for NMS-free YOLO26 differs from intended behavior), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In particular, an MRE that includes:
-Exact model + export/inference flags used (e.g., `end2end=True`, `reg_max=1`)
-A small input (one image or a tiny video clip) and the exact script used to extract features
-The exact `ultralytics` commit/version from `main`
-Any logs/prints that show the feature tensor shapes and the selection indices you’re working with

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
You’re right that the missing piece for “native” ReID in YOLO26 end-to-end is the selection index: today the top-k anchor index `idx` exists only inside `Detect.postprocess()` / `get_topk_index()` and isn’t exposed through the public prediction API, so there isn’t a supported way to retrieve those feature-map indices without a small code change to surface or stash `idx` (see the current implementation in the [Detect head reference](https://docs.ultralytics.com/reference/nn/modules/head/)).  

If you want a robust solution without relying on internal anchor/level bookkeeping, I’d recommend using the existing crop-based encoder path (it works regardless of NMS vs end-to-end because it embeds each detection crop directly); BotSORT already ships this as `ReID` in [the tracker reference](https://docs.ultralytics.com/reference/trackers/bot_sort/). Point sampling at box centers is typically too weak for identity features, so if you stay “native”, ROIAlign/box pooling over a region on higher-res features will usually be more discriminative, but it still needs a consistent mapping (or a defined pooling level policy) once you’ve exposed `idx`.
